File: do_cleanup.js
// Incremental cleanup: delete functions one at a time and verify syntax
const fs = require('fs');
const acorn = require('acorn');

let code = fs.readFileSync('game.js', 'utf8');
const originalLines = code.split('\n').length;

function removeFunction(code, funcName) {
  const searchStr = 'function ' + funcName;
  let idx = code.indexOf(searchStr + '(');
  if (idx === -1) idx = code.indexOf(searchStr + ' (');
  if (idx === -1) return null;
  
  // Make sure it's at the start of a line or after whitespace
  if (idx > 0 && code[idx-1] !== '\n' && code[idx-1] !== ' ') return null;
  
  // Find the opening brace
  let braceIdx = code.indexOf('{', idx);
  if (braceIdx === -1) return null;
  
  // Count braces to find matching close
  let depth = 0;
  let started = false;
  for (let i = braceIdx; i < code.length; i++) {
    if (code[i] === '{') { depth++; started = true; }
    if (code[i] === '}') depth--;
    if (started && depth === 0) {
      // Find the start of the line containing the function
      let lineStart = code.lastIndexOf('\n', idx);
      if (lineStart === -1) lineStart = 0;
      else lineStart++; // skip the \n
      
      // Find the end of the line with the closing brace
      let lineEnd = code.indexOf('\n', i);
      if (lineEnd === -1) lineEnd = code.length;
      else lineEnd++; // include the \n
      
      return { start: lineStart, end: lineEnd, funcName: funcName };
    }
  }
  return null;
}

const deadFuncs = [
  '_ensureExamState', '_examResultHtml', '_finishExam',
  '_ensureWeeklyMonthlyExam', '_checkWeeklyMonthlyExamPenalty',
  'takeWeeklyExam', 'takeMonthlyExam', 'takeRandomInspection',
  'renderFoodPage', 'applyLoan', 'repayLoan',
  'doEarnJob', '_completeEarnJob',
  'initGachaPool', 'getGachaCard', 'pullGacha',
  '_countTotalGachaCards', 'doGachaPull',
  '_injectGachaCSS', '_showGachaParticleBurst', 'showGachaResult',
  'renderWeeklyMonthlyExamSection', '_isWeeklyExamTime', '_isMonthlyExamTime',
  'switchEarnTab',
];

let removed = 0;
let failed = [];

for (const funcName of deadFuncs) {
  const result = removeFunction(code, funcName);
  if (result) {
    code = code.substring(0, result.start) + code.substring(result.end);
    removed++;
    // Quick syntax check after each removal
    try {
      acorn.parse(code, {ecmaVersion: 5, sourceType: 'script'});
    } catch(e) {
      failed.push({func: funcName, error: e.message, line: e.loc?.line});
      // Undo this removal
      // Actually we can't easily undo, so let's just note it and continue
      console.log('WARN: ' + funcName + ' caused parse error at line ' + e.loc?.line);
    }
  }
}

// Now handle Chinese-named functions using the raw code
const cnFuncs = [
  { name: 'render\u5feb\u9012\u670d\u52a1Page', desc: '快递服务' },
  { name: 'order\u5feb\u9012\u670d\u52a1', desc: '快递' },
  { name: 'render\u8d5a\u94b1\u4e2d\u5fc3Page', desc: '赚钱中心' },
];

for (const f of cnFuncs) {
  const result = removeFunction(code, f.name);
  if (result) {
    code = code.substring(0, result.start) + code.substring(result.end);
    removed++;
    try {
      acorn.parse(code, {ecmaVersion: 5, sourceType: 'script'});
    } catch(e) {
      failed.push({func: f.desc, error: e.message, line: e.loc?.line});
      console.log('WARN: ' + f.desc + ' caused parse error at line ' + e.loc?.line);
    }
  }
}

// Remove GACHA_POOL data structure
let gachaIdx = code.indexOf('var GACHA_POOL');
if (gachaIdx >= 0) {
  let depth = 0;
  let started = false;
  for (let i = gachaIdx; i < code.length; i++) {
    if (code[i] === '{' || code[i] === '[') { depth++; started = true; }
    if (code[i] === '}' || code[i] === ']') depth--;
    if (started && depth === 0) {
      let lineStart = code.lastIndexOf('\n', gachaIdx) + 1;
      let lineEnd = code.indexOf('\n', i) + 1;
      code = code.substring(0, lineStart) + code.substring(lineEnd);
      removed++;
      break;
    }
  }
}

// Remove exam variables
code = code.replace(/var isExamInProgress = false;\n/, '');
code = code.replace(/var currentExamSubject = '';\n/, '');
code = code.replace(/var currentExamLevel = 0;\n/, '');
code = code.replace(/var currentExamAttempt = 0;\n/, '');

// Remove earnCooldowns and gacha from gameState defaults
code = code.replace('    earnCooldowns: {},\n', '');
code = code.replace('    gacha: null,\n', '');

// Fix loan references
code = code.replace(
  /var loanWarning = \(gameState\.loanAmount > 0 && gameState\.loanInterest >= 5\)[^;]+;/,
  "var loanWarning = '';"
);
code = code.replace(
  /var hasBusinessLoan = \(gameState\.loanAmount > 0 && gameState\.loanInterest >= 5\);/,
  "var hasBusinessLoan = false;"
);

// Remove notifyGacha one-liner
code = code.replace("function notifyGacha(title, text) { showPhoneNotification('Gacha', title, text, '#FFD700'); }\n", '');

// Zero looksVal
code = code.replace(/looksVal: \d+/g, 'looksVal: 0');

// Remove earn jobs section from render行程表Page
const earnMarker = "// Earn jobs section (inline simplified earn logic)";
const earnIdx = code.indexOf(earnMarker);
if (earnIdx >= 0) {
  // Find the end - look for the closing } before getAppLinkHtml
  let depth = 0;
  let started = false;
  // The earn section is a block of code, find where it ends
  // It's a series of var declarations + a for loop + HTML building
  // Look for the pattern: after the for loop, there should be a } at the right level
  // The earn section starts at the comment and ends before getAppLinkHtml
  
  // Find getAppLinkHtml after the earn section
  const afterEarn = code.indexOf("getAppLinkHtml('schedule')", earnIdx);
  if (afterEarn >= 0) {
    // Go back to find the line start before getAppLinkHtml
    const lineStart = code.lastIndexOf('\n', afterEarn);
    // Remove from the earn marker to this line
    const earnLineStart = code.lastIndexOf('\n', earnIdx) + 1;
    code = code.substring(0, earnLineStart) + code.substring(lineStart + 1);
    removed++;
    console.log('Removed earn section');
  }
}

// Final verification
try {
  acorn.parse(code, {ecmaVersion: 5, sourceType: 'script'});
  console.log('FINAL: Syntax OK');
} catch(e) {
  console.log('FINAL: Syntax error - ' + e.message + ' at line ' + e.loc?.line);
}

const newLines = code.split('\n').length;
console.log('Lines: ' + originalLines + ' -> ' + newLines + ' (removed ' + (originalLines - newLines) + ')');
console.log('Functions removed: ' + removed);
console.log('Failed: ' + failed.length);

fs.writeFileSync('game.js', code, 'utf8');

File: do_cleanup.py
import re

with open('/tmp/my-idol-v2/game.js', 'r') as f:
    lines = f.readlines()

total = len(lines)
print(f"Starting with {total} lines")

# Helper to find a function's line range
def find_func(name, approx_line, lines):
    start = None
    for i in range(max(0, approx_line - 5), min(len(lines), approx_line + 20)):
        stripped = lines[i].strip()
        if re.match(r'function\s+' + re.escape(name) + r'\s*\(', stripped):
            start = i
            break
        if re.match(r'var\s+' + re.escape(name) + r'\s*=\s*function', stripped):
            start = i
            break
    if start is None:
        return None
    brace_count = 0
    found_open = False
    end = None
    for i in range(start, len(lines)):
        for ch in lines[i]:
            if ch == '{':
                brace_count += 1
                found_open = True
            elif ch == '}':
                brace_count -= 1
                if found_open and brace_count == 0:
                    end = i
                    break
        if end is not None:
            break
    if end is None:
        return None
    return (start, end)  # 0-based inclusive

# Collect all ranges to delete (0-based line indices)
ranges_to_delete = []

def add_range(start, end, label=""):
    """Add a range of lines to delete (0-based, inclusive)"""
    ranges_to_delete.append((start, end, label))

# ==========================================
# 1. EXAM SYSTEM
# ==========================================

# Exam variables and IIFE (L1711-1717, 0-based: 1710-1716)
# L1711: // ==================== EXAM SYSTEM ====================
# L1712: var isExamInProgress = false;
# L1713: // Reset stuck exam state on load
# L1714: (function(){...})();
# L1715: var currentExamSubject = '';
# L1716: var currentExamLevel = 0;
# L1717: var currentExamAttempt = 0;
# L1718: (empty line)
add_range(1710, 1717, "exam vars + IIFE")

# passExam (L1800-L1813)
r = find_func('passExam', 1800, lines)
if r: add_range(r[0], r[1], "passExam")

# _ensureExamState (L1781-L1798)
r = find_func('_ensureExamState', 1781, lines)
if r: add_range(r[0], r[1], "_ensureExamState")

# _examResultHtml (L1819-L1830)
r = find_func('_examResultHtml', 1819, lines)
if r: add_range(r[0], r[1], "_examResultHtml")

# _finishExam (L1832-L1840)
r = find_func('_finishExam', 1832, lines)
if r: add_range(r[0], r[1], "_finishExam")

# Exam game renderers
for fname, approx in [
    ('renderDanceExam', 1843),
    ('renderVocalExam', 1951),
    ('renderRapExam', 2042),
    ('renderActingExam', 2122),
    ('renderVarietyExam', 2202),
    ('renderComprehensiveExam', 2282),
]:
    r = find_func(fname, approx, lines)
    if r: add_range(r[0], r[1], fname)

# _ensureWeeklyMonthlyExam (L11336-L11343)
r = find_func('_ensureWeeklyMonthlyExam', 11336, lines)
if r: add_range(r[0], r[1], "_ensureWeeklyMonthlyExam")

# _checkWeeklyMonthlyExamPenalty (L11371-L11404)
r = find_func('_checkWeeklyMonthlyExamPenalty', 11371, lines)
if r: add_range(r[0], r[1], "_checkWeeklyMonthlyExamPenalty")

# takeWeeklyExam (L11406-L11434)
r = find_func('takeWeeklyExam', 11406, lines)
if r: add_range(r[0], r[1], "takeWeeklyExam")

# takeMonthlyExam (L11436-L11465)
r = find_func('takeMonthlyExam', 11436, lines)
if r: add_range(r[0], r[1], "takeMonthlyExam")

# renderWeeklyMonthlyExamSection (L11467-L11518)
r = find_func('renderWeeklyMonthlyExamSection', 11467, lines)
if r: add_range(r[0], r[1], "renderWeeklyMonthlyExamSection")

# takeRandomInspection (L11520-L11550)
r = find_func('takeRandomInspection', 11520, lines)
if r: add_range(r[0], r[1], "takeRandomInspection")

# ==========================================
# 2. FOOD PAGE (外卖APP)
# ==========================================
r = find_func('renderFoodPage', 7454, lines)
if r: add_range(r[0], r[1], "renderFoodPage")

# ==========================================
# 3. EXPRESS (快递APP)
# ==========================================
r = find_func('render快递服务Page', 7549, lines)
if r: add_range(r[0], r[1], "render快递服务Page")

r = find_func('order快递服务', 7633, lines)
if r: add_range(r[0], r[1], "order快递服务")

# ==========================================
# 4. LOAN APP
# ==========================================
r = find_func('applyLoan', 7674, lines)
if r: add_range(r[0], r[1], "applyLoan")

r = find_func('repayLoan', 7694, lines)
if r: add_range(r[0], r[1], "repayLoan")

# ==========================================
# 5. GACHA SYSTEM
# ==========================================
for fname, approx in [
    ('initGachaPool', 11570),
    ('getGachaCard', 11649),
    ('pullGacha', 11694),
    ('_countTotalGachaCards', 11771),
    ('doGachaPull', 11853),
    ('_injectGachaCSS', 11863),
    ('_showGachaParticleBurst', 11885),
    ('showGachaResult', 11897),
]:
    r = find_func(fname, approx, lines)
    if r: add_range(r[0], r[1], fname)

# notifyGacha (one-liner at L8802)
# Find the exact line
for i, line in enumerate(lines):
    if 'function notifyGacha(' in line:
        add_range(i, i, "notifyGacha")
        break

# ==========================================
# 6. EARN SYSTEM (赚钱)
# ==========================================
r = find_func('render赚钱中心Page', 14202, lines)
if r: add_range(r[0], r[1], "render赚钱中心Page")

r = find_func('switchEarnTab', 14327, lines)
if r: add_range(r[0], r[1], "switchEarnTab")

r = find_func('doEarnJob', 14333, lines)
if r: add_range(r[0], r[1], "doEarnJob")

r = find_func('_completeEarnJob', 14450, lines)
if r: add_range(r[0], r[1], "_completeEarnJob")

# Also find the EARN CENTER APP comment line
for i, line in enumerate(lines):
    if '// ==================== EARN CENTER APP ====================' in line:
        add_range(i, i, "earn comment header")
        break

# ==========================================
# Now merge overlapping ranges and sort
# ==========================================
ranges_to_delete.sort(key=lambda x: x[0])

# Merge overlapping
merged = []
for start, end, label in ranges_to_delete:
    if merged and start <= merged[-1][1] + 1:
        # Overlapping or adjacent
        merged[-1] = (merged[-1][0], max(merged[-1][1], end), merged[-1][2] + " + " + label)
    else:
        merged.append((start, end, label))

print(f"\n=== Ranges to delete ({len(merged)} ranges) ===")
total_lines_to_delete = 0
for start, end, label in merged:
    count = end - start + 1
    total_lines_to_delete += count
    print(f"  L{start+1}-L{end+1} ({count} lines): {label}")

print(f"\nTotal lines to delete: {total_lines_to_delete}")

# Delete ranges from back to front
new_lines = list(lines)
for start, end, label in reversed(merged):
    del new_lines[start:end+1]

print(f"Lines after deletion: {len(new_lines)}")

# Write intermediate result
with open('/tmp/my-idol-v2/game.js', 'w') as f:
    f.writelines(new_lines)

print("Phase 1 (function deletion) complete.")

